---
title: "How to use the immport.api package"
author: "Gabriel Rosenfeld"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{How to use the immport.api package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
# immport.api

## Background

The [Immunology Database and Analysis Portal (ImmPort)](http://www.immport.org/immport-open/public/home/home) has been developed under the Bioinformatics Integration Support Contract (BISC) Phase II by the Northrop Grumman Information Technology Health Solutions team for the [National Institutes of Health (NIH), National Institute of Allergy and Infectious Diseases (NIAID), Division of Allergy, Immunology, and Transplantation (DAIT)](https://www.niaid.nih.gov/about/dait).

This database provides [public access to immunology research data](http://www.immport.org/immport-open/public/home/studySearch) generated by DAIT-funded research programs spanning basic research to clinical trials designed to increase our understanding of the causes and mechanisms that lead to the development of allergic and immunologic diseases.  The immunology research data covers a multitude of data types from genetic, genomic, proteomic, elisa, elispot, immunoassay, flow cytometry, and CyTOF as well as clinical data such as lab tests, adverse events, treatments, assessments, etc.

Recently, ImmPort has developed an [application programming interface (API)](http://docs.immport.org/#API/DataQueryAPI/dataqueryapi/) to the public shared data.  This R package was developed as a wrapper over the API service to provide an easy and convenient way of accessing the data directly through R.

## Setting up the immport.api wrapper

### What is required before installing immport.api package?

Before installing the immport.api R package. You must register an account with [ImmPort](https://immport-user-admin.niaid.nih.gov:8443/registrationuser/registration) and agree to abide by the user agreement.  Once you have registered, you will receive a user name and password to download data from the data repository.  This ImmPort user name and password are the same user name and password that immport.api requires to programmatically access the data.

### Installation of the immport.api R package

Install the immport.api R package using the code below:

```{r eval = FALSE}
library(devtools)
devtools::install_github("grosenf1/immport.api")
library(immport.api)
```

## How to use the immport.api R package

It is recommended to first look through the [API documentation](http://docs.immport.org/#API/DataQueryAPI/dataqueryapi/) on the ImmPort website to familiarize yourself with the functionality provided.  The API allows access to a variety of data sources and also provides filters to focus in on particular set of data of interest.  For instance, you could look at elisa data involving patients with certain demographic characteristics (age, gender, or race).

### Short cut to the data type and filters that are used by the API

immport.api provides several short cut functions to look at the available data types and filters that can be sent to the API.

```{r eval = FALSE}
# Generate a data frame of all the data types the API can access
dt <- show_data_types()
View(dt)

# Generate a data frame of all the filters the API can use to subset data
f <- show_filters()
View(f)
```

### Generating a token with user name and password

Before submitting a query to the API, you must generate a unique token using your ImmPort user name and password.  This token expires every few minutes so needs to be regenerated periodically.  If you notice that the API does not return a result, try regenerating your token.

```{r eval = FALSE}
# Generate your unique API token
t <- gen_token(INSERT_USERNAME, INSERT_PASSWORD)
```

### Example query from the API documentation

Here is an example of how you would do the query in the API documentation using the immport.api package: "https://api.immport.org/data/query/result/elisa?studyAccession=SDY2,SDY4"

```{r eval = FALSE}
# Perform the API query
q <- data_query(token = t, data_type = "elisa", filter = list(studyAccession = c("SDY2", "SDY4")))

# Process the API result into a data frame
r <- proc_response(q)

View(r)
```

### Example query to obtain clinical ELISA data from subjects over the age of 85

```{r eval = FALSE}
# Perform the API query and process into data frame in one step
r <- data_query(token = t, data_type = "elisa", filter = list(clinical = "Y", maxSubjectAgeGte = 85))

View(r)
```
